pandas/core/groupby/generic.py

```diff
@@ -1083,7 +1083,7 @@ def _cython_agg_blocks(
                         result = type(block.values)._from_sequence(
                             result.ravel(), dtype=block.values.dtype
                         )
-                    except ValueError:
+                    except (ValueError, TypeError):
```

Comment:
Can you share the backtrace for this? @jbrockmendel did a lot of cleanup of Exception catching in groupby and in the spirit of being explicit I'm a little hesitant to add more back in. The error *might* even be that a TypeError is getting thrown instead of a ValueError

Reply:
Here's the full traceback:

```python
In [1]: import pandas as pd                                                                                                                                            

In [2]: df = pd.DataFrame({"a": [1, 1, 2, 2], "b": [1, 2, 1, 2]}, dtype="Int64")                                                                                       

In [3]: df.groupby("a").mean()                                                                                                                                         
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/pandas/pandas/core/arrays/integer.py in safe_cast(values, dtype, copy)
    155     try:
--> 156         return values.astype(dtype, casting="safe", copy=copy)
    157     except TypeError:

TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'safe'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-3-f32730bc62ca> in <module>
----> 1 df.groupby("a").mean()

~/pandas/pandas/core/groupby/groupby.py in mean(self, numeric_only)
   1224             "mean",
   1225             alt=lambda x, axis: Series(x).mean(numeric_only=numeric_only),
-> 1226             numeric_only=numeric_only,
   1227         )
   1228 

~/pandas/pandas/core/groupby/generic.py in _cython_agg_general(self, how, alt, numeric_only, min_count)
    993     ) -> DataFrame:
    994         agg_blocks, agg_items = self._cython_agg_blocks(
--> 995             how, alt=alt, numeric_only=numeric_only, min_count=min_count
    996         )
    997         return self._wrap_agged_blocks(agg_blocks, items=agg_items)

~/pandas/pandas/core/groupby/generic.py in _cython_agg_blocks(self, how, alt, numeric_only, min_count)
   1082                         # Cast back if feasible
   1083                         result = type(block.values)._from_sequence(
-> 1084                             result.ravel(), dtype=block.values.dtype
   1085                         )
   1086                     except ValueError:

~/pandas/pandas/core/arrays/integer.py in _from_sequence(cls, scalars, dtype, copy)
    356     @classmethod
    357     def _from_sequence(cls, scalars, dtype=None, copy: bool = False) -> "IntegerArray":
--> 358         return integer_array(scalars, dtype=dtype, copy=copy)
    359 
    360     @classmethod

~/pandas/pandas/core/arrays/integer.py in integer_array(values, dtype, copy)
    142     TypeError if incompatible types
    143     """
--> 144     values, mask = coerce_to_array(values, dtype=dtype, copy=copy)
    145     return IntegerArray(values, mask)
    146 

~/pandas/pandas/core/arrays/integer.py in coerce_to_array(values, dtype, mask, copy)
    259         values = safe_cast(values, dtype, copy=False)
    260     else:
--> 261         values = safe_cast(values, dtype, copy=False)
    262 
    263     return values, mask

~/pandas/pandas/core/arrays/integer.py in safe_cast(values, dtype, copy)
    162 
    163         raise TypeError(
--> 164             f"cannot safely cast non-equivalent {values.dtype} to {np.dtype(dtype)}"
    165         )
    166 

TypeError: cannot safely cast non-equivalent float64 to int64
```

To be honest I'm not sure if this should be a `TypeError` or `ValueError`. `numpy` doesn't even seem consistent about this kind of thing:

```python
In [8]: np.array(["a", "b"]).astype(int)                                                                                     
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-5f3d565c8777> in <module>
----> 1 np.array(["a", "b"]).astype(int)

ValueError: invalid literal for int() with base 10: 'a'

In [9]: np.array(["a", "b"]).astype(int, casting="safe")                                                                     
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-9-0a3ea49fa702> in <module>
----> 1 np.array(["a", "b"]).astype(int, casting="safe")

TypeError: Cannot cast array from dtype('<U1') to dtype('int64') according to the rule 'safe'
```

Reply:
hmm, what do we raise `ValueError` on in construction of an EA?

Reply:
I think a case could be made that this "should" be a `ValueError`. It's `safe_cast` that actually raises but it evidently raises because of the _value_ and not the type:

```python
In [9]: safe_cast(np.array([1.0], dtype=float), dtype="int", copy=False)                                                                                               
Out[9]: array([1])

In [10]: safe_cast(np.array([1.5], dtype=float), dtype="int", copy=False)                                                                                              
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/pandas/pandas/core/arrays/integer.py in safe_cast(values, dtype, copy)
    155     try:
--> 156         return values.astype(dtype, casting="safe", copy=copy)
    157     except TypeError:

TypeError: Cannot cast array from dtype('float64') to dtype('int64') according to the rule 'safe'

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
<ipython-input-10-c8293cc0cca1> in <module>
----> 1 safe_cast(np.array([1.5], dtype=float), dtype="int", copy=False)

~/pandas/pandas/core/arrays/integer.py in safe_cast(values, dtype, copy)
    162 
    163         raise TypeError(
--> 164             f"cannot safely cast non-equivalent {values.dtype} to {np.dtype(dtype)}"
    165         )
    166 

TypeError: cannot safely cast non-equivalent float64 to int64
```

I could try switching it to a `ValueError` and see if anything breaks as a result?